### PR TITLE
C++ convert_numpy_inputs, _as_ndarray, _is_ndarray

### PR DESCRIPTION
**Pull request by jonathanchu33**
This should precede the merge of a C++ version of `canonicalize_function_inputs`

### CODE REVIEW

**Comment by jonathanchu33** [comment]
Benchmarks over 3 batches of 10 trials each

Minimum
defun_matmul_2_by_2_CPU: 151 -> 154.5 us
defun_matmul_2_by_2_CPU_async: 80.5 -> 81 us
defun_matmul_2_by_2_with_signature_CPU: 206 -> 206 us

Mean
defun_matmul_2_by_2_CPU: 159.5 -> 158 us
defun_matmul_2_by_2_CPU_async: 81.5 -> 83 us
defun_matmul_2_by_2_with_signature_CPU: 216.5 -> 214 us

Median
defun_matmul_2_by_2_CPU: 159.5 -> 158 us
defun_matmul_2_by_2_CPU_async: 81.5 -> 83 us
defun_matmul_2_by_2_with_signature_CPU: 213 -> 215 us

Inconclusive

File: tensorflow/python/eager/function.cc
Comment (kkimdev) [inline tensorflow/python/eager/function.cc:135]: prefer static_cast over c-style cast https://google.github.io/styleguide/cppguide.html#Casting

File: tensorflow/python/eager/function.cc
Comment (kkimdev) [inline tensorflow/python/eager/function.cc:130]: This is a very long condition, can you factor out as a function?

**Review by kkimdev — CHANGES_REQUESTED** [review]

**Comment by kkimdev** [comment]
Offline discussion(for record): migrating the counter itself saves little but additionally migrating `FrequentTracingDetector` saves ~1us so let's also include that.  Then we can also avoid using a lock(https://github.com/tensorflow/tensorflow/blob/1d5a16dee74d2fab18cea44654c194cec4a45161/tensorflow/python/eager/def_function.py#L92) because we can rely on Python GIL.

```
MicroBenchmarks.benchmark_cpp_counter
time               : 0.19 us
runs_per_second    : 5274818

MicroBenchmarks.benchmark_python_counter
time               : 0.28 us
runs_per_second    : 3619124

MicroBenchmarks.benchmark_freq_detector
time               : 1.30 us
runs_per_second    : 770743
```

File: tensorflow/python/eager/function.cc
Comment (jonathanchu33) [reply tensorflow/python/eager/function.cc]: Some of them caused seg faults when not defined as here. But after code revisions the declarations here are consistent anyways

File: tensorflow/python/eager/function.cc
Comment (jonathanchu33) [reply tensorflow/python/eager/function.cc]: Done

File: tensorflow/python/eager/function.cc
Comment (jonathanchu33) [reply tensorflow/python/eager/function.cc]: Done

File: tensorflow/python/eager/function.cc
Comment (jonathanchu33) [reply tensorflow/python/eager/function.cc]: Fixed. Tried searching this up but didn't find anything without probing too deeply. I was getting errors when trying to return a py::handle though.

File: tensorflow/python/eager/function.cc
Comment (jonathanchu33) [reply tensorflow/python/eager/function.cc]: Obsolete

File: tensorflow/python/eager/function.cc
Comment (jonathanchu33) [reply tensorflow/python/eager/function.cc:120]: Couldn't find anything.. :|

File: tensorflow/python/eager/function.cc
Comment (jonathanchu33) [reply tensorflow/python/eager/function.cc]: Obsolete

File: tensorflow/python/eager/function.cc
Comment (jonathanchu33) [reply tensorflow/python/eager/function.cc]: I haven't changed it, but the syntax is slightly more readable now. I didn't factor it out originally because it was around half of IsNdarray, and using it there would have invoked another function call. I'm also not sure how to group these conditions under a name. Otherwise, I've left a comment indicating that it is equivalent to the IsNdarray == True case.
Thoughts? Can change if it still seems too dense.

**Comment by jonathanchu33** [comment]
Updated Benchmarks over 3 batches of 10 trials each

Minimum
defun_matmul_2_by_2_CPU: 157 -> 154.5 us
defun_matmul_2_by_2_CPU_async: 79 -> 79 us
defun_matmul_2_by_2_with_signature_CPU: 207 -> 210.5 us

Mean
defun_matmul_2_by_2_CPU: 167 -> 163 us
defun_matmul_2_by_2_CPU_async: 81.5 -> 81.5 us
defun_matmul_2_by_2_with_signature_CPU: 234 -> 229.5 us

Median
defun_matmul_2_by_2_CPU: 166.5 -> 161 us
defun_matmul_2_by_2_CPU_async: 81.5 -> 81.5 us
defun_matmul_2_by_2_with_signature_CPU: 221 -> 221 us

Inconclusive

**Review by sun51 — COMMENTED** [review]
thanks, overall LGTM,  please wait for kkimdev @ for final approval

**Comment by jonathanchu33** [comment]
Yes, the benchmarks in the last comment are updated @sun51

**Review by sun51 — DISMISSED** [review]

**Review by kkimdev — DISMISSED** [review]

**Review by kkimdev — APPROVED** [review]

**Comment by sun51** [comment]
this pr is still not merged yet.

**Comment by gbaned** [comment]
@jonathanchu33  Can you please resolve conflicts? Thanks!

**Comment by gbaned** [comment]
@jonathanchu33  Any update on this PR? Please. Thanks!

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by gbaned** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!

**Pull request closed without merging**
